man/man1/csf.1
.TH CSF 1NEMO "26 May 2024"

.SH "NAME"
csf \- copy structured file

.SH "SYNOPSIS"
\fBcsf in=\fP\fIin-file\fP \fBout=\fP\fIout-file\fP

.SH "DESCRIPTION"
\fIcsf\fP copies a NEMO structured file.  By using the \fBitem\fP and
\fBselect\fP keyword items can be selected by name and/or number.  Note
that items can currently only be selected from the top level, items
within a set can not be selected this way; they are copied as a whole. 
This is likely to be improved at a later stage.  Automatic
data-conversion can also be done.  By default all items are copied, and
no data conversion is applied. 
.PP
An overview of the top level items that are present in a structured
file can be obtained with the following Unix pipe:
.EX

    % tsf in_file | grep "^[a-z,A-Z]" | grep -v "^tes" | awk '{print $2}'
.EE
.PP
Although \fIstructured files\fP can always be read when they were created
on a byte swapped machine, \fIcsf\fP can also be used to create a file
in which the bytes are not swapped.
.PP
As a convenience option, the keyword \fBblocksize=\fP can be used to forced
a fixed blocksize to copy a file "as is", but it will bypass the structured
file I/O so should be functionally the same as a unix \fIcp(1)\fP.

.SH "PARAMETERS"
.so man1/parameters
.TP 24
\fBin=\fP\fIin-file\fP
Input file, must be in \fIfilestruct(5NEMO)\fP format
[no default].
.TP
\fBout=\fP\fIout-file\fP
Output file, will be in \fIfilestruct(5NEMO)\fP format
[no default].
.TP
\fBitem=\fP\fIitem-list\fP
List of top level items, separated by comma's or spaces, to be copied.
If none are provided, all items are selected. Note that the 
\fIitem-list\fP will be case sensitive.
[default: -blank-].
.TP
\fBselect=\fP\fIselect-list\fP
List of numbers (1 being the first) that are select for 
copying, after the above mentioned item-name selection
(from the \fBitem=\fP keyword) has been applied.
If none are provided, all are copied.
[default: -blank-].
.TP
\fBconvert=\fP\fIconvert-list\fP
List of conversions to make, separated by comma's or spaces. Each conversion
is of the format \fIfrom\fP2\fIto\fP, where both the \fIfrom\fP and
\fIto\fP part are any of: \fBd\fP (double), \fBf\fP (float), \fBh\fP 
(half precision) ,
\fBl\fP (long), \fBi\fP (int) or \fBs\fP (short). 
[Default: -blank-].
.TP
\fBblocksize=\fP
If selected, this is the blocksize, in bytes, by which a file gets copied
bypassing NEMO's structured file I/O. A value of 512 is ok. Default: 0 (not used)
.TP
\fBheadline=\fP
If supplied, add this string of random verbiage to the data stream. [Default: none]

.SH "CAVEATS"
Some of the conversions in \fBconvert=\fP
are not supported or are in danger of loosing more than just precision.

.SH "EXAMPLES"
An image with too much data-history can be cut by only copying the \fIImage\fP
items:
.EX

   % csf in=map1.ccd out=map2.ccd select=Image
.EE

.SH "CAVEATS"
Items can only be selected (\fBselect=\fP) from the top level.

.SH "SEE ALSO"
cp(1), tsf(1NEMO), rsf(1NEMO)

.SH "AUTHOR"
Peter Teuben, Joshua E. Barnes

.SH "FILES"
.nf
.ta +1.5i
~/src/kernel/io  	csf.c
.fi

.SH "HISTORY"
.nf
.ta +1.25i +4.5i
xx-xxx-87	V1.0: created	JEB
9-dec-89	V1.1 added select keyword	PJT
10-oct-90	V1.3 select= is now item=: select= used for counting	PJT
4-mar-91	V1.4 added data conversion option via convert=    	PJT
13-feb-92	documentation added	PJT
26-mar-95	V1.5 fixed selection bug and clarified doc	PJT
12-dec-09	V1.6 added support for half precision (halfp) type	PJT
13-aug-2022	V1.7 add headline=	PJT
19-apr-2023	V1.8 add blocksize=	PJT
.fi


src/kernel/io/csf.c
/*
 * CSF: copy structured file.
 *
 *      xx-xxx-87   V1.0 Original version - Josh Barnes
 *       9-dec-89   V1.1 added select= keyword	PJT
 *	 4-apr-90   V1.2 fixed 'select=all' bug	PJT
 *	 9-oct-90   V1.3 select= now item= ; must now be matched by case too
 *			 select= now counts which ordinal ones
 *	 5-mar-91   V1.4 allow type switching
 *	18-jul-91       a    last printf is now dprintf(0,
 *	19-feb-92       b    deleted two unused variables
 *	19-aug-92       c    removed lint complaint about free()
 *      20-feb-94       d    ANSI
 *	26-mar-95   V1.5  fixed counting bug in select= and item=, 1 is now
 *			  the first item in all item= selected (as meant)
 *			a fixed NULL vs. 0 warning
 *      11-dec-09   V1.6  experimenting with half precision 
 *      19-apr-23   V1.8  allow raw "cp" style copy using blocksize=
 */

#include <stdinc.h>
#include <getparam.h>
#include <extstring.h>
#include <ctype.h>
#include <filestruct.h>

string defv[] = {
    "in=???\n		Input file",
    "out=???\n		Output file",
    "item=\n		Top level selection items [default: all]",
    "select=\n          Selection numbers (1...) [default: all]",
    "convert=\n		Conversion options {d2f,f2d,i2f,f2i,d2i,i2d,h2d,d2h}",
    "blocksize=0\n      If selected, use raw I/O with this blocksize, bypassing structured I/O",
    "headline=\n        Add additional headline",
    "VERSION=1.8a\n	18-dec-2023 PJT",
    NULL,
};

string usage = "Copy a binary structured file";

#define  MAXSEL 512

extern string *burststring(string, string);

local bool check(string *, string);


void nemo_main(void)
{
    stream instr, outstr, nullstr;
    string item, *items;
    string *tags, *sels, *cvt;
    int    i, nsel, select[MAXSEL], cntrd, cntwr;
    int blocksize = getiparam("blocksize");
    //bool all = TRUE;

    if (blocksize > 0) {
      dprintf(1,"raw I/O using blocksize=%d\n",blocksize);
      char *data = allocate(blocksize);
      size_t ndata, nsum = 0;
      instr = stropen(getparam("in"), "r");
      outstr = stropen(getparam("out"), "w");
      while ((ndata=fread(data, 1, blocksize, instr))) {
	nsum += ndata;
	fwrite(data, 1, ndata, outstr);
      }
      dprintf(1,"Wrote %ld\n",nsum);
      return;
    }

    instr = stropen(getparam("in"), "r");
    nullstr = stropen("/dev/null","w+");        /* kludge: force write */
    item = getparam("item");
    sels = burststring(getparam("select"),", ");
    nsel = xstrlen(sels, sizeof(string)) - 1;
    for (i=0; i<nsel; i++) {		/* note: array should be sorted */
        select[i] = atoi(sels[i]);
	dprintf(1,"#selected #%d\n",select[i]);
	if (select[i]<1)
            error("%d: bad number for select=, must be > 0",select[i]);
        if (i>0 && (select[i]<select[i-1]))
            error("%d: bad number for select=, array must be sorted",select[i]);
    }
    cvt = burststring(getparam("convert"),", ");

    if (*item == 0) {
        //all = TRUE;
	items = (string *) allocate(sizeof(string));    /* PPAP */
	items[0] = NULL;
    } else {
        //all = FALSE;
        items = burststring(item,", ");
        if (items==NULL)
            error("error parsing item=%s\n",item);
        dprintf(1,"selected: ");
        for (i=0; items[i]!=NULL; i++)
            dprintf(1," %s",items[i]);
        dprintf(1,"\n\n");
    }
    outstr = stropen(getparam("out"), "w");
    if (hasvalue("headline"))
      put_string(outstr, HeadlineTag, getparam("headline"));
    cntrd = 0;      /* keep track of items read */
    cntwr = 0;      /* and written */
    while ((tags = list_tags(instr)) != NULL) {
        if (check(items, *tags)) {
            cntrd++;
            dprintf(1," %d: %s",cntrd,*tags);
	    if (nsel==0 || (nsel>0 && select[cntwr] == cntrd)) {
               copy_item_cvt(outstr, instr, *tags, cvt);
               cntwr++;
               dprintf(1," copied\n");
	    } else {
	       copy_item(nullstr, instr, *tags);
               dprintf(1," skipped\n");
            }
        } else {
            dprintf(1," *: %s",*tags);        	
	    copy_item(nullstr, instr, *tags);
            dprintf(1," skipped\n");
        }
	free((char *)*tags);
	free((char *)tags);
        if (nsel>0 && cntwr >= nsel)     /* early bailout ? */
            break;
    }
    dprintf(0,"On top-level: %d items  read, %d items written\n",cntrd,cntwr);
    strclose(instr);
    strclose(outstr);
}

/* check if name is in list of items - with minimum match in the items list */

local bool check(
string *items,      /* list if items, user wants to have selected */
string name         /* the set name, as found in file */
) {
    int i;
    char lname[32];

    if (items[0] == NULL) return TRUE;		/* designate all */
    strcpy(lname,name);
#if 0
    for (i=0; lname[i]!=NULL; i++)		/* convert to lower case */
        if (isupper(lname[i]))
            lname[i] = tolower(lname[i]);
#endif

    for (i=0; items[i]!=NULL; i++)
        if (streq(items[i],lname))
            return TRUE;
    return FALSE;
}
